### PR TITLE
Make node symbols templatable in the graphs

### PR DESCRIPTION
**Pull request by algmyr**
No-op change styling wise for now. I suspect there will be room for bikeshedding the default style after this.

Hopefully I'm not stepping on @zummenix's toes too much here after the great idea in #3260 to just use a separate template for the elided node. :sweat_smile:

Some stuff is a bit odd still, like what template to even use for the elided node? Commit template for the root it happens to be associated with? Just make it not a template and let it be a string?

Still need to update docs.

Hopefully just dropping the symbol configs I added recently is fine, especially considering there hasn't been a release with them included.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [X] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

### CODE REVIEW

**Comment by zummenix** [comment]
> Hopefully I'm not stepping on @zummenix's toes too much here after the great idea in https://github.com/martinvonz/jj/issues/3260 to just use a separate template for the elided node. 😅

Not at all. I appreciate you working on this, thank you!

I skimmed through the changes. Nice work! Seems fine to me, but I'll let owners properly review them.

> Some stuff is a bit odd still, like what template to even use for the elided node? Commit template for the root it happens to be associated with? Just make it not a template and let it be a string?

I wondered almost the same questions. I was also wondering should there be different templates for different kinds of graphs (log, op log, obslog)? And users would be able to override them in their config, like for example:
```toml
[templates]
graph-log = 'my_fancy_graph_log'
graph-op-log = 'my_simple_graph_log'
graph-obslog = 'my_simple_graph_log'
```

File: cli/src/cli_util.rs
Comment (yuja) [inline cli/src/cli_util.rs:2320]: You'll probably need to use `ui.new_formatter(buffer)` to get colorized output.

File: cli/src/config-schema.json
Comment (yuja) [inline cli/src/config-schema.json:126]: nit: other templates are defined in `[templates]` section, so I think these node templates should be named something like `templates.log_graph_node`.

File: lib/src/settings.rs
Comment (yuja) [inline lib/src/settings.rs:250]: nit: it would be nice if we can move them to `config/templates.toml`, but there are `"ascii"` stuff.

Do we still need them to be the graph-style dependent by default?

File: cli/src/config-schema.json
Comment (algmyr) [reply cli/src/config-schema.json:126]: I'm surprised that `templates` isn't in the config schema, is that intentional?

Comment (yuja) [reply cli/src/config-schema.json:126]: Indeed. I believe it's an oversight. The `revsets`/`revset-aliases` pair is documented.

File: cli/src/cli_util.rs
Comment (algmyr) [reply cli/src/cli_util.rs:2320]: Done. Though speaking of colors. Is there any support for users to specify colors in templates, or is it just through labels that gets applied when template functions/properties are used?

Comment (yuja) [reply cli/src/cli_util.rs:2320]: Only labels, but there's `label()` template function to apply whatever you want.

File: lib/src/settings.rs
Comment (algmyr) [reply lib/src/settings.rs:250]: Having sensible defaults that follow from picking the graph style seems important, though I agree having to put defaults here is suboptimal.

I guess one option could be to split the template in 2 to have a dedicated ascii variant in the config. Or there could be an `ascii` property available in the template (though then we are back to extending template languages again, which is a bit iffy atm).

File: cli/src/cli_util.rs
Comment (algmyr) [reply cli/src/cli_util.rs:2320]: Ah, neat. I totally overlooked that, tested and coloring does work.

File: cli/src/commands/log.rs
Comment (algmyr) [inline cli/src/commands/log.rs:125]: Ok, in light on the comments about labels and coloring having this not be a plain string is preferable. I'll make it an empty `GenericTemplateLanguage` for the time being. LMK if this is fine.

Comment (yuja) [reply cli/src/commands/log.rs:126]: Sounds good to me. Hopefully we can merge the elided template with the commit one later.

File: cli/src/commands/operation.rs
Comment (yuja) [inline cli/src/commands/operation.rs:180]: nit: no need to construct `language` twice.

It also applies to the commit template (and is more important since the underlying cache can be shared), but there's no helper function to do so.

File: cli/src/cli_util.rs
Comment (yuja) [inline cli/src/cli_util.rs:2314]: nit: `template`, not `template_language`

File: lib/src/settings.rs
Comment (algmyr) [reply lib/src/settings.rs:250]: Ok, let's leave things as is for clean-up later. When we have the ability to extend languages we can add an elided property along with an ascii property, and then the default can be moved to the toml.

File: cli/src/cli_util.rs
Comment (algmyr) [reply cli/src/cli_util.rs]: Done.

**Comment by algmyr** [comment]
Another open question. It's totally feasible to include the elided revision text in the template, and it seems some people have been eager to have the ability to not have that. Is it worth making the default something like this?
```
'"◌  " ++ label("elided", "(elided revisions)")'
```

File: cli/src/commands/log.rs
Comment (martinvonz) [reply cli/src/commands/log.rs]: > Hopefully we can merge the elided template with the commit one later.

I actually like the idea of splitting them up because there's almost no overlap in what properties they share. Elided nodes are not associated with any particular commit.

**Comment by martinvonz** [comment]
> '"◌  " ++ label("elided", "(elided revisions)")'

Then we would have to split that up after reading it from the config (or maybe after rendering it), right? Because the node is passed separately from the text to the graph renderer.

**Comment by algmyr** [comment]
> Then we would have to split that up after reading it from the config (or maybe after rendering it), right? Because the node is passed separately from the text to the graph renderer.

One option is to just write a newline in the graph, but it feels a bit hacky and maybe the two spaces is not always the right spacing?
```diff
         let elided_key = (elided_target, true);
         let real_key = (elided_key.0.clone(), false);
         let edges = [Edge::Direct(real_key)];
-        let mut buffer = vec![];
-        with_content_format.write_graph_text(
-            ui.new_formatter(&mut buffer).as_mut(),
-            |formatter| writeln!(formatter.labeled("elided"), "(elided revisions)"),
-            || graph.width(&elided_key, &edges),
-        )?;
         let node_symbol = format_template(ui, &(), elided_node_template.as_ref());
         graph.add_node(
             &elided_key,
             &edges,
             &node_symbol,
-            &String::from_utf8_lossy(&buffer),
+            &"\n",
         )?;
     }
 }

```

**Comment by martinvonz** [comment]
> is not always the right spacing?

Right, there can be lines to the right of the elided node, like this:

```
◌ │  (elided revisions)
├─╯
```

**Comment by algmyr** [comment]
Gotcha, let's just leave things as is then. I guess there could be yet another template introduced in the future if someone really wants it/

File: cli/src/config-schema.json
Comment (algmyr) [reply cli/src/config-schema.json:126]: Initial issue fixed, leaving config schema as is since existing templates aren't documented currently anyway.

File: cli/src/commands/operation.rs
Comment (algmyr) [reply cli/src/commands/operation.rs:180]: It might be worth refactoring this at some point, but idk what a really nice way of doing this is. The straightforward options are to either construct language twice, or to let the constructed language live for longer than it needs to.

Doing something like this would work, but feels pretty icky imo
```
let (tmpl1, tmpl2) = {
    let language = ...;
    (parse_template(template_string1), parse_template(template_string2))
};
```
and would require refactoring `parse_commit_template` into something like a `get_commit_template_language` and using parse template separately.

Comment (yuja) [reply cli/src/commands/operation.rs:180]: It could be
```
let tmpl1;
let tmpl2;
{
    ...
    tmpl1 = parse(..);
    tmpl2 = parse(..);
}
```
(if you prefer.)

I grouped the `language` and parsing variables just because I generally prefer shorter name and scope. It's not super important.

> would require refactoring parse_commit_template into something like a get_commit_template_language and using parse template separately.

Yep. https://github.com/martinvonz/jj/pull/3319/commits/52a400e1581555851766367bff35749a96acc976

Comment (algmyr) [reply cli/src/commands/operation.rs]: Oh neat, I completely forgot you can separate declaration and definition like that for non-mut stuff. Stacked another commit with this change on top, can squash if needed later.

File: CHANGELOG.md
Comment (yuja) [inline CHANGELOG.md:26]: nit: maybe better to name them as `log_graph_node`, `op_log_graph_node`, and `log_graph_node_elided`. The text parts are `log` and `op_log` respectively.

Comment (algmyr) [reply CHANGELOG.md:26]: Maybe graph could even be left out if the emphasis is on the command
```
log_node
log_node_elided
op_log_node
```

Comment (yuja) [reply CHANGELOG.md:26]: I like the former (assuming `_elided` will eventually be combined into `log_node`.)

>  I guess one possible confusion with the first one is that the commit and elided ones also apply to the obslog.

Yep, but that's consistent with the `templates.log` text template. We might want separate `templates.obslog*` definitions, but I don't know.

Comment (algmyr) [reply CHANGELOG.md]: Done.

File: cli/tests/test_obslog_command.rs
Comment (martinvonz) [inline cli/tests/test_obslog_command.rs:159]: We can delete this line, right?

File: cli/tests/test_obslog_command.rs
Comment (martinvonz) [inline cli/tests/test_obslog_command.rs:160]: nit: don't include a case for the root since it's not used

**Review by martinvonz — COMMENTED** [review]
Looks good to me, thanks, but I'll let @yuja give the final approval since he's already done a few rounds of reviews

File: lib/src/settings.rs
Comment (yuja) [inline lib/src/settings.rs:249]: remove these noop `if(immutable, x, x)`

File: cli/src/commands/log.rs
Comment (yuja) [inline cli/src/commands/log.rs:120]: nit: `template_string` can also be moved in this block.

File: cli/src/commands/obslog.rs
Comment (yuja) [inline cli/src/commands/obslog.rs:82]: nit: ditto. `template_string` can also be moved in this block.

File: docs/config.md
Comment (yuja) [inline docs/config.md:233]: nit: insert blank line after the header

File: cli/src/cli_util.rs
Comment (yuja) [inline cli/src/cli_util.rs:912]: Suppose the second patch will be squashed, do you want to include this change in your patch, or rebase on top of mine?

**Review by yuja — APPROVED** [review]
Yes, looks good, thanks!

Two questions:
* Will you merge this soon? or rebase on atop #3332?
* Do you like the idea of rendering elided node as `None`? https://github.com/martinvonz/jj/pull/3319#issuecomment-2011011160

(To be clear, I don't intend to block this PR. I can reorder my PRs even if we agree with `Option<Commit>` idea.)

File: lib/src/settings.rs
Comment (algmyr) [reply lib/src/settings.rs:249]: Done.

File: cli/src/commands/log.rs
Comment (algmyr) [reply cli/src/commands/log.rs:120]: Done.

File: cli/src/commands/obslog.rs
Comment (algmyr) [reply cli/src/commands/obslog.rs:82]: Done.

File: docs/config.md
Comment (algmyr) [reply docs/config.md:233]: Done.

File: cli/tests/test_obslog_command.rs
Comment (algmyr) [reply cli/tests/test_obslog_command.rs:159]: Done.

File: cli/tests/test_obslog_command.rs
Comment (algmyr) [reply cli/tests/test_obslog_command.rs:160]: Done.

File: cli/src/cli_util.rs
Comment (algmyr) [reply cli/src/cli_util.rs:912]: I'm ready to merge this now. Hopefully my impl is close enough to yours that the conflicts are minimal. :)

**Pull request merged**
